platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -157,57 +161,52 @@ protected void onEmptyOrAbsentSourceFile(final Path path) throws IOException {
-            final int lastChunkWithDataIndex = totalNumberOfChunks - firstChunkWithDataIndex - 1;
-
-            // copy the first chunk
-            final ByteBuffer transferBuffer = initOrGetTransferBuffer();
-            // we need to make sure that the chunk is written in full.
-            // If a value is absent, the list element will have IMPERMISSIBLE_VALUE
-            fillBufferWithZeroes(transferBuffer);
-            transferBuffer.position(minValidIndexInChunk * Long.BYTES);
-            MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
-            transferBuffer.flip();
-            // writing the full chunk, all values before minValidIndexInChunk are zeroes
-            MerkleDbFileUtils.completelyWrite(tempFileCHannel, transferBuffer, 0);
-            chunkList.set(firstChunkWithDataIndex, 0L);
-
-            // copy everything except for the first chunk and the last chunk
-            final int numberOfFullChunks = totalNumberOfChunks - firstChunkWithDataIndex - 2;
-            if (numberOfFullChunks > 0) {
-                final long bytesToTransfer = (long) numberOfFullChunks * memoryChunkSize;
-                final long bytesTransferred = MerkleDbFileUtils.completelyTransferFrom(
-                        tempFileCHannel, fileChannel, memoryChunkSize, bytesToTransfer);
-                if (bytesTransferred != bytesToTransfer) {
-                    throw new IOException("Failed to read long list chunks, expected=" + bytesToTransfer + " actual="
-                            + bytesTransferred);
-                }
+            this.tempFileChannel = rf.getChannel();
+            super.readBodyFromFileChannelOnInit(sourceFileName, fileChannel);
+        } finally {
+            if (this.tempFileChannel != null && this.tempFileChannel.isOpen()) {
+                this.tempFileChannel.close();
             }
+        }
+    }
 
-            // copy the last chunk
-            transferBuffer.clear();
-            MerkleDbFileUtils.completelyRead(fileChannel, transferBuffer);
-            transferBuffer.flip();
-            MerkleDbFileUtils.completelyWrite(
-                    tempFileCHannel, transferBuffer, (long) lastChunkWithDataIndex * memoryChunkSize);
+    /** {@inheritDoc} */
+    @Override
+    protected Long readChunkData(FileChannel fileChannel, int chunkIndex, int startIndex, int endIndex)
+            throws IOException {
+        final ByteBuffer transferBuffer = initOrGetTransferBuffer();
+        fillBufferWithZeroes(transferBuffer);
 
-            for (int i = firstChunkWithDataIndex + 1; i < totalNumberOfChunks; i++) {
-                chunkList.set(i, (long) (i - firstChunkWithDataIndex) * memoryChunkSize);
-            }
+        readDataIntoBuffer(fileChannel, chunkIndex, startIndex, endIndex, transferBuffer);
+
+        final int firstChunkWithDataIndex = toIntExact(minValidIndex.get() / numLongsPerChunk);
+        return (long) (chunkIndex - firstChunkWithDataIndex) * memoryChunkSize;
```

Comment:
Here is the problem. In `LongListDisk` a chunk is not just a long (offset), but a long + currentFileChannel. Chunk content is available in the channel by offset, which is chunk ID (long). That's why the chunk must be written to `currentFileChannel` in the end of this method, otherwise chunk content is just lost.